File: leetcode/p1291/Solution.java
class Solution {
    public List<Integer> sequentialDigits(int low, int high) {
        int numCount = getNumCount(low);
        int base = getSeqBase(numCount);
        List<Integer> ans = new ArrayList<>();
        int cur = base;
        while (cur <= high) {
            if (cur >= low) {
                ans.add(cur);
            }
            cur = getNextSeq(cur);
        }
        return ans;
    }
    
    private int getNumCount(int num) {
        int temp = num;
        int count = 0;
        while (temp > 0) {
            temp /= 10;
            count++;
        }
        return count;
    }
    
    private int getSeqBase(int count) {
        int seq = 0;
        for (int i = 1; i <= count; ++i) {
            seq = seq * 10 + i;
        }
        return seq;
    }
    
    private int getNextSeq(int seq) {
        int count = getNumCount(seq);
        int inc = 0;
        for(int i = 1; i <= count; ++i) {
            inc = inc * 10 + 1;
        }
        int next = seq + inc;
        
        return (next % 10 == 0) ? getSeqBase(count + 1) : next;
    }
}

File: leetcode/p1291/Solution.kt
class Solution {
    fun sequentialDigits(low: Int, high: Int): List<Int> {
        val list = mutableListOf<Int>()
        val map = mutableMapOf<Int, Int>()
        var temp = 1
        for (i in 2..9) {
            temp = temp * 10 + i
            
            map[i] = temp
        }
        
        if (low > map[9]!!) return list
        
        val count = getNumCount(low)
    
        var num = map[count]!!
        while (num < low) {
            num = getNextSeq(num, map)
        }
        while (num <= high) {
            list.add(num)
            if (num == 123456789) break
            
            num = getNextSeq(num, map)
        }
        return list
    }
    
    private fun getNumCount(num: Int): Int {
        var cur = num
        var count = 0
        while (cur != 0) {
            count++
            cur /= 10
        }
        return count
    }
    
    private fun getNextSeq(num: Int, map: Map<Int, Int>): Int {
        var count = getNumCount(num)
        var curNum = num
        if (curNum % 10 == 9) {
            return map[count + 1]!!
        }
        
        count = getNumCount(curNum)
        var next = curNum
        var pow = 1
        for (i in 1..count) {
            next += pow
            pow *= 10
        }
        return next
    }
    
}
